### PR TITLE
Change FORMAT/ENCODE syntax for Kafka/Kinesis sink

### LINKED ISSUE #1350: Document: feat(sink): support new syntax `FORMAT ... ENCODE ...` similar to source
This issue tracks the documentation update needed for the merged PR #12556.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/12556
Source PR Merged At: 2023-10-04T04:49:05Z

### PR DESCRIPTION
**Pull request by emile-00**
## Info

- **Description**

  Update the type/format syntax for Kafka/Kinesis sink. Pulsar sink syntax will be updated with the Pulsar sink doc PR.

- **Notes**

  [ Include any supplementary context or references here. ]

- **Related code PR**

  https://github.com/risingwavelabs/risingwave/pull/12556

- **Related doc issue**
  
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/1350

## For reviewers

- **Preview**

 - Kafka: https://pr-1371.d2fbku9n2b6wde.amplifyapp.com/docs/dev/create-sink-kafka/
 - Kinesis: https://pr-1371.d2fbku9n2b6wde.amplifyapp.com/docs/dev/sink-to-aws-kinesis/
 - CREATE SINK: https://pr-1371.d2fbku9n2b6wde.amplifyapp.com/docs/dev/sql-create-sink/

- **Key points**

  [ Parts that may need revision or extra consideration. ]

## Before merging

- [ ] I have checked the doc site preview, and the updated parts look good.

- [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`CharlieSYH`, `emile-00`, & `hengm3467`).

### CODE REVIEW

**Comment by xiangjinwu** [comment]
* The format is named `plain` rather than `append-only`. This is aligned with source.
* The support for `avro` or `protobuf` is not complete yet. The only accepted `encode` now is `json`.
* Note that we now have 2 pairs of parentheses, and it may be necessary to clarify which options belong to which pair. Due to the lack of support of `avro` or `protobuf`, the only option that goes to the second pair is `force_append_only`.
* Shall we update the syntax reference page ([link](https://docs.risingwave.com/docs/current/sql-create-sink/)) as well? We can list `format ... encode ...` as optional and say they are only available for kafka/kinesis/pulsar now.

**Comment by emile-00** [comment]
@xiangjinwu I've edited the pages according to your review! If you have the time, please take another look. Regarding the 3rd bullet point, what about the `primary_key` parameter? Would that be in the first set of parentheses under the `WITH` options?

File: docs/guides/create-sink-kafka.md [hunk 1]
```diff
@@ -26,7 +26,11 @@ CREATE SINK [ IF NOT EXISTS ] sink_name
 WITH (
    connector='kafka',
    connector_parameter = 'value', ...
-);
+)
+FORMAT data_format ENCODE data_encode [ (
+    force_append_only = 'value'
```
Comment (xiangjinwu) [inline docs/guides/create-sink-kafka.md]: Either be fully generic `format_parameter = 'value'` or fully specific `force_append_only = [true | false]`

File: docs/guides/create-sink-kafka.md [hunk 2]
```diff
@@ -47,11 +51,17 @@ All `WITH` options are required except `force_append_only` and `primary_key`.
...
 |`primary_key`| The primary keys of the sink. Use ',' to delimit the primary key columns. If the external sink has its own primary key, this field should not be specified.|
 |`properties.client.id`|Optional. Client ID associated with the Kafka client. |
 
+## Sink parameters
+
+|Field|Notes|
+|-----|-----|
+|data_format| Data format. Allowed formats:<ul><li> `PLAIN`: Output data with insert operations.</li><li> `DEBEZIUM`: Output change data capture (CDC) log in Debezium format.</li><li> `UPSERT`: Output data as a changelog stream. `primary_key` must be specified in this case. </li></ul> To learn about when to define the primary key if creating an `UPSERT` sink, see the [Overview](/data-delivery.md).|
+|data_encode| Data encode. Supported encode: `JSON`. |
+|force_append_only| If `true`, forces the sink to be `PLAIN`, even if it cannot be.|
```
Comment (xiangjinwu) [inline docs/guides/create-sink-kafka.md]: nit: `PLAIN`: also known as append-only

**Review by xiangjinwu — COMMENTED** [review]
> Regarding the 3rd bullet point, what about the `primary_key` parameter? Would that be in the first set of parentheses under the `WITH` options?

Yes it is in the first part.

Some other issues regarding `primary_key`, orthogonal to this PR:
> If the external sink has its own primary key, this field should not be specified.

This is no longer true ([link](https://risingwave-labs.slack.com/archives/C03L8DNMDDW/p1693879379703809?thread_ts=1693879275.310829&cid=C03L8DNMDDW)), and even the opposite. To keep it simple, we can just be specific to kafka and kinesis here:
* kafka: optional for `PLAIN` but required for `UPSERT` / `DEBEZIUM`.
* kinesis: always required (recent [bug fix](https://github.com/risingwavelabs/risingwave/pull/12810))

File: docs/guides/sink-to-aws-kinesis.md [hunk 3]
```diff
@@ -37,14 +41,15 @@ WITH (
 |aws.credentials.session_token |Optional. The session token associated with the temporary security credentials. |
 |aws.credentials.role.arn |Optional. The Amazon Resource Name (ARN) of the role to assume.|
 |aws.credentials.role.external_id|Optional. The [external id](https://aws.amazon.com/blogs/security/how-to-use-external-id-when-granting-access-to-your-aws-resources/) used to authorize access to third-party resources. |
+|primary_key| Conditional. The primary keys of the sink. Use ',' to delimit the primary key columns. Primary keys are optional when creating a `PLAIN` sink but required for `UPSERT` and `DEBEZIUM` sinks.|
```
Comment (xiangjinwu) [inline docs/guides/sink-to-aws-kinesis.md]: For kinesis,`primary_key` is always required.

**Review by xiangjinwu — APPROVED** [review]

**Review by xiangjinwu — COMMENTED** [review]
Btw this is also part of v1.3, whose snapshot is already taken.

@CharlieSYH @hengm3467 Could you help expedite this?

File: docs/guides/create-sink-kafka.md [hunk 4]
```diff
@@ -51,6 +55,14 @@ All `WITH` options are required unless explicitly mentioned as optional.
 |`force_append_only`| Optional. If `true`, forces the sink to be `append-only`, even if it cannot be.|
```
Comment (xiangjinwu) [inline docs/guides/create-sink-kafka.md]: `type` has been deprecated in favor of `format`. `force_append_only` is also moved.

File: docs/guides/create-sink-kafka.md [hunk 5]
```diff
@@ -47,10 +51,17 @@ All `WITH` options are required unless explicitly mentioned as optional.
...
 |`properties.bootstrap.server`|Address of the Kafka broker. Format: `‘ip:port’`. If there are multiple brokers, separate them with commas. |
 |`topic`|Address of the Kafka topic. One sink can only correspond to one topic.|
-|`type`|Data format. Allowed formats:<ul><li> `append-only`: Output data with insert operations.</li><li> `debezium`: Output change data capture (CDC) log in Debezium format.</li><li> `upsert`: Output data as a changelog stream. `primary_key` must be specified in this case. </li></ul> To learn about when to define the primary key if creating an `upsert` sink, see the [Overview](/data-delivery.md).|
-|`force_append_only`| Optional. If `true`, forces the sink to be `append-only`, even if it cannot be.|
+|`format`|Data format. Allowed formats:<ul><li> `append-only`: Output data with insert operations.</li><li> `debezium`: Output change data capture (CDC) log in Debezium format.</li><li> `upsert`: Output data as a changelog stream. `primary_key` must be specified in this case. </li></ul> To learn about when to define the primary key if creating an `upsert` sink, see the [Overview](/data-delivery.md).|
```
Comment (xiangjinwu) [inline docs/guides/create-sink-kafka.md]: The `data_format` clause is already described below. There is no `format` option in the first part.

File: docs/guides/create-sink-kafka.md (hunk 4 above)
Comment (emile-00) [reply docs/guides/create-sink-kafka.md]: Thanks for catching that! The merge with the main branch made me miss these details

**Comment by hengm3467** [comment]
> Btw this is also part of v1.3, whose snapshot is already taken.
> 
> @CharlieSYH @hengm3467 Could you help expedite this?

Sorry. I missed this mention.

**Pull request merged**